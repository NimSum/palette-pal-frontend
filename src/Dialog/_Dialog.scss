.dialog-overlay {
  align-items: center;
  background: rgba(0, 0, 0, 0.8);
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: center;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  z-index: 3;
}

.popup {
  background: white;
  padding: 25px 30px 25px 30px;
  border-radius: 10px;
  font-size: 20px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 300px;
  position: relative;

  select > option:first-child{
    display: none;
  }

  .fa-times {
    color: darkred;
    position: absolute;
    top: 12px;
    right: 10px;

    &:hover {
        color: rgb(230, 67, 67);
        cursor: pointer;
      }
  }

  h3 {
    font-weight: 400;
    font-size: 24px;
    color: $title-color;
    margin-bottom: 15px;
    align-self: center;
  }

  ::placeholder {
    font-style: italic;
    font-size: 13px;
    color: rgb(176, 176, 176);
  }

  .dialog-divider {
    display: flex;
    width: 100%;
    justify-content: space-between;
    align-items: center;

    hr {
      flex: 3;
      height: 1px;
      opacity: .4;
    }

    p {
      margin: 0 5px;
      color: rgb(150, 150, 150);
      font-size: 15px;
    }
  }

  .dialog-msg {
    font-size: 16px;
    margin-bottom: 15px;
  }

  .palette-preview {
    width: 100%;
    height: 16px;
    display: flex;
    margin-bottom: 20px;

    .preview-color {
      flex: 1;
      height: 100%;
    }
  }

  select,
  input {
    margin: 8px 0 12px 0;
    padding-left: 5px;
    height: 28px;
    width: 100%;
    border-radius: 5px;
    font-size: 14px;
  }

  label {
    font-size: 14px;
    margin-top: 10px;
    align-self: flex-start;
  }
  
  .dialog-btns {
    display: flex;
    justify-content: flex-end;
    text-decoration: none;
    width: 100%;
  }
  
  .dialog-btn {
    margin-top: 20px;
    margin-left: 10px;
    height: 30px;
    width: 74px;
    border: none;
    border-radius: 5px;
    color: white;
    font-size: 16px;
    padding-bottom: 3px;
    background: $btn-color;
    cursor: pointer;
    
    &:hover {
      background: $btn-hover;
    }
    
    &.cancel-btn {
      color: darkred;
      border: 1px solid darkred;
      background: white;
      
      &:hover {
        color: rgb(230, 67, 67);
        border: 1px solid rgb(230, 67, 67);
      }
    }
  }
}
  