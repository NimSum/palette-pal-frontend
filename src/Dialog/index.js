import React, { Component } from 'react';

class Dialog extends Component {
  constructor(props) {
    super(props);

    this.state = {
      projects: [],
      paletteName: '',
      projectID: 0,
      newProject: ''
    }
  }

  componentDidMount() {
    window.removeEventListener('keydown', this.props.refreshUnheldColors);

    fetch('http://localhost:3005/api/v1/projects')
      .then(res => res.json())
    .then(projects => this.setState({projects}))
  }

  componentWillUnmount() {
    window.addEventListener('keydown', this.props.refreshUnheldColors);
  }

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  }
  
  handleClick = () => {
    const { paletteName, projectID, newProject } = this.state;
    const data = { paletteName, projectID };

    if (this.state.newProject) {
      fetch('http://localhost:30001/api/v1/projects', {
        method: 'POST',
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          name: newProject
        })
      })
        .then(res => data = { paletteName, projectID: res.body.id })
    }
    
    this.props.primaryAction(data);
    this.setState({
      paletteName: '',
      projectID: 0,
      newProject: ''
    })
  }

  render() {
    const colors = Object.values(this.props.colors);
    const colorDivs = colors.map(color => <div className="preview-color" key={color} style={{ backgroundColor: color }}></div>);
    const projectOptions = this.state.projects.map(p => {
      return <option key={p.id} value={p.id}>{p.name}</option>
    });

    return (
      <div className="dialog-overlay">
        <div className="popup">
          <i className="fas fa-times" onClick={this.props.closeDialog}></i>
          <h3>Save New Palette</h3>
          <input className="dropdown-input palette-name-input" name="paletteName" placeholder="Enter Palette Name..." onChange={this.handleChange}></input>
          <div className="palette-preview">
            {colorDivs}
          </div>
          <label htmlFor="projectID">Choose A Project</label>
          <select className="dropdown-input project-input" name="projectID" onChange={this.handleChange}>
            {projectOptions}
          </select>
          <div className="dialog-divider"><hr /><p>OR</p><hr /></div>
          <label htmlFor="newProject">Create New Project</label>
          <input className="project-input" name="newProject" placeholder="Enter Project Name..." onChange={this.handleChange}></input>
          <div className="dialog-btns">
            <button className="dialog-btn cancel-btn" type="button" onClick={this.props.closeDialog} >
              Cancel
            </button>
            <button className="dialog-btn" type="button" onClick={this.handleClick}>
              Save
            </button>
          </div>
        </div>
      </div>
    );
  }
}

export default Dialog;
