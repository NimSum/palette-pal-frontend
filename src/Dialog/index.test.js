import React from 'react';
import Dialog from './index';
import { shallow } from 'enzyme';

describe('Dialog', () => {
  let wrapper;
  let instance;
  window.addEventListener = jest.fn();
  window.removeEventListener = jest.fn();
  const mockCloseDialog = jest.fn();
  const mockRefreshUnheldColors = jest.fn();
  const mockUpdateProjectData = jest.fn();
  const mockPrimaryAction = jest.fn();
  const mockColors = [
    '#ffffff',
    '#000000',
    '#232323',
    '#454545',
    '#765456'
  ]
  const mockData = [{
    id: 1,
    name: "Uncategorized",
    palettes: [
      {
        name: 'Bob',
        id: 3,
        colors: {
          color_1: '#ffffff',
          color_2: '#000000',
          color_3: '#232323',
          color_4: '#454545',
          color_5: '#765456'
        }
      }
    ]
  }]

  beforeEach(() => {
    wrapper = shallow(<Dialog
      title="Save New Palette"
      type="newPalette"
      closeDialog={mockCloseDialog}
      refreshUnheldColors={mockRefreshUnheldColors}
      updateProjectData={mockUpdateProjectData}
      primaryAction={mockPrimaryAction}
      colors={mockColors}
      data={mockData}
    />, { disableLifecycleMethods: true });

    instance = wrapper.instance();

    jest.spyOn(instance, 'sendFormData');
    jest.spyOn(instance, 'getPaletteFields');
    jest.spyOn(instance, 'handleChange');
    jest.spyOn(instance, 'handleClick');
    jest.spyOn(instance, 'conveyResult');
    jest.spyOn(instance, 'setDefaultProjectOption');
  })

  it('should match component Dialog snapshot when the dialog is for saving a new palette', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match component Dialog snapshot when the dialog is for saving a new project', () => {
    wrapper = shallow(<Dialog
      title="Create New Project"
      type="newProject"
      closeDialog={mockCloseDialog}
      primaryAction={mockPrimaryAction}
    />, { disableLifecycleMethods: true });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match component Dialog snapshot when the dialog is for logging a user in', () => {
    wrapper = shallow(<Dialog
      title="Login"
      type="login"
      closeDialog={mockCloseDialog}
      primaryAction={mockPrimaryAction}
    />, { disableLifecycleMethods: true });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match component Dialog snapshot when the dialog is for signing a user in', () => {
    wrapper = shallow(<Dialog
      title="Sign In"
      type="signin"
      closeDialog={mockCloseDialog}
      primaryAction={mockPrimaryAction}
    />, { disableLifecycleMethods: true });

    expect(wrapper).toMatchSnapshot();
  });

  it('should have the expected default state', () => {
    expect(wrapper.state()).toEqual({
      palette_name: '',
      project_name: '',
      project_id: '',
      user_name: '',
      password: '',
      email: '',
      showConf: false,
      error: ''
    })
  })

  it('should remove a window event listnener and call setDefaultProjectOption on mount', () => {
    instance.componentDidMount();

    expect(instance.setDefaultProjectOption).toHaveBeenCalled();

    expect(window.removeEventListener).toHaveBeenCalled();
  })

  it('should set a window event listnener when it unmounts', () => {
    instance.componentWillUnmount();

    expect(window.addEventListener).toHaveBeenCalled();
  })

  it('should update project_id in state when setDefaultProjectOoption is called', () => {
    expect(wrapper.state('project_id')).toEqual('');

    wrapper.data = [{id: 1}]

    instance.setDefaultProjectOption();

    expect(wrapper.state('project_id')).toEqual(1);
  })

  it('should update the project_name state when handleChange is invoked and the new project input has a value', () => {
    const mockEvent = { target: { name: 'project_name', value: 'test' } }
    expect(wrapper.state('project_name')).toEqual('');

    instance.handleChange(mockEvent);

    expect(wrapper.state('project_name')).toEqual('test');
  });

  it('should update the palette_name state when handleChange is invoked and the palette name input has a value', () => {
    const mockEvent = { target: { name: 'palette_name', value: 'test' } }

    expect(wrapper.state('palette_name')).toEqual('');

    instance.handleChange(mockEvent);

    expect(wrapper.state('palette_name')).toEqual('test');
  });

  it('should update the project_id state when handleChange is invoked when the project dropdown changes', () => {
    const mockEvent = { target: { name: 'project_id', value: 3 } }

    expect(wrapper.state('project_id')).toEqual('');

    instance.handleChange(mockEvent);

    expect(wrapper.state('project_id')).toEqual(3);
  });

  it('should call handleChange method when the new item name input field\'s value changes', () => {
    jest.spyOn(instance, 'handleChange');
    const mockEvent = { target: { name: 'palette_name', value: 'called' } }

    expect(wrapper.state('palette_name')).toEqual('');

    wrapper.find('.name-input').simulate('change', mockEvent);

    expect(wrapper.state('palette_name')).toEqual('called');
  })

  it('should call handleChange method when the select field for project changes', () => {
    const mockEvent = { target: { name: 'project_id', value: 5 } }
    jest.spyOn(instance, 'handleChange');

    expect(wrapper.state('project_id')).toEqual('');

    wrapper.find('.existing-project').simulate('change', mockEvent);

    expect(wrapper.state('project_id')).toEqual(5);
  })

  it('should call handleChange method when the input for a new project name changes', () => {
    const mockEvent = { target: { name: 'project_name', value: 'howdy' } }
    jest.spyOn(instance, 'handleChange');

    expect(wrapper.state('project_name')).toEqual('');

    wrapper.find('.new-project-name').simulate('change', mockEvent);

    expect(wrapper.state('project_name')).toEqual('howdy');
  })

  it('should call the closeDialog method when the cancel button is clicked', () => {
    wrapper.find('.cancel-btn').simulate('click');

    expect(mockCloseDialog).toHaveBeenCalled();
  })

  it('should call the closeDialog method when the close button is clicked', () => {
    wrapper.find('.fa-times').simulate('click');

    expect(mockCloseDialog).toHaveBeenCalled();
  })

  it('should reset the default state when handleClick is called', async () => {
    const mockEvent = { target: { name: 'project_name', value: 'howdy' } }
    jest.spyOn(instance, 'handleClick');
    jest.spyOn(instance, 'handleChange');

    expect(wrapper.state('project_name')).toEqual('');

    wrapper.find('.new-project-name').simulate('change', mockEvent);
    
    expect(wrapper.state('project_name')).toEqual('howdy');

    await wrapper.find('.save-btn').simulate('click');

    setTimeout(() => expect(wrapper.state('project_name')).toEqual(''), 1000);
  })

  it('should call handleClick method when the user hits the save button', () => {
    jest.spyOn(instance, 'handleClick');
    jest.spyOn(instance, 'handleChange');

    expect(wrapper.state('error')).toEqual('');

    wrapper.find('.save-btn').simulate('click');

    expect(wrapper.state('error')).not.toBe('');
  })

  it('should call the primaryAction method when sendFormData is invoked', () => {
    instance.sendFormData();

    expect(mockPrimaryAction).toHaveBeenCalled();
  })

  it('should call the sendFormData method when handleClick is invoked and all required fields are filled', async () => {
    const mockEvent1 = { target: { name: 'palette_name', value: 'test pal' } }
    const mockEvent2 = { target: { name: 'project_name', value: 'test proj' } }

    wrapper.find('.name-input').simulate('change', mockEvent1);
    wrapper.find('.new-project-name').simulate('change', mockEvent2);

    expect(wrapper.state('palette_name')).toEqual('test pal');
    expect(wrapper.state('project_name')).toEqual('test proj');

    await wrapper.find('.save-btn').simulate('click');

    await expect(instance.sendFormData).toHaveBeenCalled();
  })

})