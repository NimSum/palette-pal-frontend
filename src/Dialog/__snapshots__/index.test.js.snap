// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog shoud match component Dialog snapshot 1`] = `
<div
  className="dialog-overlay"
>
  <div
    className="popup"
  >
    <i
      className="fas fa-times"
    />
    <h3 />
    <input
      className="dropdown-input name-input"
      name="projectName"
      onChange={[Function]}
      placeholder="Enter project name..."
    />
    <div
      className="dialog-btns"
    >
      <button
        className="dialog-btn cancel-btn"
        type="button"
      >
        Cancel
      </button>
      <button
        className="dialog-btn"
        onClick={[Function]}
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
